ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/services/grpc/GrpcCommandSubmissionService.scala

```diff
@@ -66,19 +70,29 @@ class GrpcCommandSubmissionService(
         .value(
           metrics.daml.commands.validation,
           validator.validate(
-            request,
+            requestWithSubmissionId,
             currentLedgerTime(),
             currentUtcTime(),
             maxDeduplicationTime(),
           )(errorLogger),
         )
         .fold(
-          t => Future.failed(ValidationLogger.logFailure(request, t)),
+          t => Future.failed(ValidationLogger.logFailure(requestWithSubmissionId, t)),
           service.submit(_).map(_ => Empty.defaultInstance),
         ),
     )
   }
 
   override def bindService(): ServerServiceDefinition =
     CommandSubmissionServiceGrpc.bindService(this, executionContext)
+
+  private def generateSubmissionIdIfEmpty(request: ApiSubmitRequest): ApiSubmitRequest = {
+    if (request.commands.exists(_.submissionId.isEmpty)) {
+      val commandsWithSubmissionId =
+        request.commands.map(_.copy(submissionId = submissionIdGenerator.generate()))
```

Comment:
You may simplify this with lenses:
```scala
  request.update(_.commands.submissionId := submissionIdGenerator.generate())
```

Reply:
Nice, I didn't know ScalaPB has lenses

Reply:
0d962b0a5dcc4fbeccafbd48fdaaecc1a69e7553